scripts/util/doc-builder.js

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env node
+
+var path = require('path');
+var fs = require('fs');
+var yuidoc = require('./yuidoc');
+var NodeGit = require('nodegit');
+var util = require('util');
+var fsExtra = require('fs-extra');
+
+function DocBuilder(options) {
+  this.options = util._extend({
+    masterBranchName: 'master',
+    docsBranchName: 'gh-pages',
+    docsDirName: 'docs',
+    versionsDirName: 'versions',
+    srcDirName: 'src',
+    rmSrcDir: false,
+    commitDocs: false,
+    checkoutStrategy: ( NodeGit.Checkout.STRATEGY.FORCE + NodeGit.Checkout.STRATEGY.DONT_WRITE_INDEX ),
+  }, options);
+
+  if (!this.options.themeDir) {
+    this.options.themeDir = path.join(this.options.docsDirName, 'yuidoc', 'theme');
+  }
+
+  this.options.apiDirPath = path.join(this.options.docsDirName, this.options.versionsDirName);
+}
+
+DocBuilder.prototype.ensureDocsBranchExists = function () {
+  console.info('info:', 'Ensuring \'' + this.options.docsBranchName + '\' branch exists');
+  return NodeGit.Repository.open('.').then(repo => {
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(references => {
```

Comment:
I suggested encapsulating your logic into more functions instead of using the giant tree, and creating scopes instead of globals when you had to persist state. Which means a combination of the two.